### ISSUE TITLE
Issue #5439: New Check: Enforce or deny a blank line after the left curly brace

### ISSUE THREAD

**Issue by pbludov** [checkstyle/checkstyle#5439]
A rule to enforce or forbid a blank line after the left (opening) curly brace.

By default this check validates presence of a blank line after the left curly brace.
To configure this check to validate absence of any blank lines at the top of the code block,
set "blankLine" parameter to `false`.

No validation is performed for empty code blocks, since this check may conflict with
`BeforeRightCurly` check if they are configured in the opposite way.

To enforce a blank line after the left curly brace:
&lt;module name="AfterLeftCurly"&gt;
&nbsp;&nbsp;&lt;property name="blankLine" value="true"/&gt;
&lt;/module&gt;

```
interface Foo {

    public int method1();
    public int method2();

}
```

To forbid blank lines after the left curly brace:
&lt;module name="AfterLeftCurly"&gt;
&nbsp;&nbsp;&lt;property name="blankLine" value="false"/&gt;
&lt;/module&gt;

```
interface Foo {
    public int method1();
    public int method2();
}
```

**Comment by romani** [comment]
Please provide here whole design description, as only this issue be referenced I release notes.
Sorry for bureaucracy. But it will help to see inconsistency in design and name as we will look at at as all users will do

**Comment by pbludov** [comment]
There are several nuances here:

* Check should (optionally?) ignore empty blocks (code block with some whitespace between).

```java
class One {} // ok, no violation even if a blank line is required

class Two {
} // ok, no violation even if a blank line is required

class Three {

} // ok, a blank line is here

class Four {
  /* Comment */ // violation: a blank line is required
}
```

* A style may require the blank line for some tokens, but forbid for other. For example, Checkstyle itself uses sparse types, but compact methods:

```java
class Outer {

  class Inner {

    int method(int size) {
      int result = 0;
      for (int index = 0; index < size; ++index) {
        if (index % 3 == 0) {
          result *= 2;
        }
      }
      return result;
   }

    void method2() {
    }

    void method3() {

    } // ok, empty method

    int method4() {

      return 42;  // violation, no blank line expected
    }

  }

} 
```

The advantage of this style is that looking at the closing brackets at the end of the file, you can understand whether the method or the method of the nested class has ended here.

So we need four checks:
* **BlankLineAfterLeftCurly** (possible **EmptyLineAfterLeftCurly** to match **EmptyLineSeparator**)
  with default tokens = { Annotation, Class, Enum, Interface, Record }
* **NoBlankLineAfterLeftCurly**
  with default tokens = { method/ctor, for, if, try, ... }
* **BlankLineBeforeRightCurly**
  with default tokens = { Annotation, Class, Enum, Interface, Record }
* **NoBlankLineBeforeRightCurly**
  with default tokens = { method/ctor, for, if, try, ... }

@romani @rnveach @strkkk @nick-mancuso please take a look.

**Comment by romani** [comment]
Looks like no energy here to define design, 

Some implementation is at https://github.com/checkstyle/checkstyle/pull/11525 and can be reused if  someone come to this.

**Comment by romani** [comment]
will be covered by https://github.com/checkstyle/checkstyle/issues/17393
